File: sdk/unreal/Source/ChillObservability/Private/ChillJson.cpp
// Copyright Chill Contributors. Licensed under the Apache License, Version 2.0.

#include "ChillJson.h"

namespace ChillJson
{
	FString Quote(const FString& Value)
	{
		FString Result;
		Result.Reserve(Value.Len() + 8);
		Result.AppendChar(TEXT('"'));
		for (int32 Index = 0; Index < Value.Len(); ++Index)
		{
			const TCHAR Character = Value[Index];
			switch (Character)
			{
			case TEXT('"'):
				Result.Append(TEXT("\\\""));
				break;
			case TEXT('\\'):
				Result.Append(TEXT("\\\\"));
				break;
			case TEXT('\b'):
				Result.Append(TEXT("\\b"));
				break;
			case TEXT('\f'):
				Result.Append(TEXT("\\f"));
				break;
			case TEXT('\n'):
				Result.Append(TEXT("\\n"));
				break;
			case TEXT('\r'):
				Result.Append(TEXT("\\r"));
				break;
			case TEXT('\t'):
				Result.Append(TEXT("\\t"));
				break;
			default:
				if (Character < 0x20)
				{
					Result.Append(FString::Printf(TEXT("\\u%04x"), static_cast<int32>(Character)));
				}
				else
				{
					Result.AppendChar(Character);
				}
				break;
			}
		}
		Result.AppendChar(TEXT('"'));
		return Result;
	}

	FString StringValue(const FString& Value)
	{
		return TEXT("{\"stringValue\":") + Quote(Value) + TEXT("}");
	}

	FString BooleanValue(bool Value)
	{
		return FString(TEXT("{\"boolValue\":")) + (Value ? TEXT("true") : TEXT("false")) + TEXT("}");
	}

	FString IntegerValue(int64 Value)
	{
		return IntegerValue(FString::Printf(TEXT("%lld"), Value));
	}

	FString IntegerValue(const FString& Value)
	{
		return TEXT("{\"intValue\":") + Quote(Value) + TEXT("}");
	}

	FString DoubleValue(double Value)
	{
		// Shortest round-trippable form, matching the "R" format the other SDKs use.
		FString Text = FString::Printf(TEXT("%.17g"), Value);
		for (int32 Precision = 1; Precision < 17; ++Precision)
		{
			const FString Candidate = FString::Printf(TEXT("%.*g"), Precision, Value);
			if (FCString::Atod(*Candidate) == Value)
			{
				Text = Candidate;
				break;
			}
		}
		return TEXT("{\"doubleValue\":") + Text + TEXT("}");
	}

	FString StringArrayValue(const TArray<FString>& Values)
	{
		FString Result = TEXT("{\"arrayValue\":{\"values\":[");
		for (int32 Index = 0; Index < Values.Num(); ++Index)
		{
			if (Index > 0)
			{
				Result.AppendChar(TEXT(','));
			}
			Result.Append(StringValue(Values[Index]));
		}
		Result.Append(TEXT("]}}"));
		return Result;
	}

	FString Attribute(const FString& Key, const FString& AnyValueJson)
	{
		return TEXT("{\"key\":") + Quote(Key) + TEXT(",\"value\":") + AnyValueJson + TEXT("}");
	}

	static FString BooleanArrayValue(const TArray<bool>& Values)
	{
		FString Result = TEXT("{\"arrayValue\":{\"values\":[");
		for (int32 Index = 0; Index < Values.Num(); ++Index)
		{
			if (Index > 0)
			{
				Result.AppendChar(TEXT(','));
			}
			Result.Append(BooleanValue(Values[Index]));
		}
		Result.Append(TEXT("]}}"));
		return Result;
	}

	static FString NumberArrayValue(const TArray<double>& Values)
	{
		FString Result = TEXT("{\"arrayValue\":{\"values\":[");
		for (int32 Index = 0; Index < Values.Num(); ++Index)
		{
			if (Index > 0)
			{
				Result.AppendChar(TEXT(','));
			}
			Result.Append(DoubleValue(Values[Index]));
		}
		Result.Append(TEXT("]}}"));
		return Result;
	}

	FString AnyValue(const FChillValue& Value)
	{
		switch (Value.Kind)
		{
		case EChillValueKind::String:
			return StringValue(Value.StringValue);
		case EChillValueKind::Boolean:
			return BooleanValue(Value.BooleanValue);
		case EChillValueKind::Number:
			return DoubleValue(Value.NumberValue);
		case EChillValueKind::Strings:
			return StringArrayValue(Value.StringValues);
		case EChillValueKind::Booleans:
			return BooleanArrayValue(Value.BooleanValues);
		case EChillValueKind::Numbers:
			return NumberArrayValue(Value.NumberValues);
		}
		return StringValue(FString());
	}

	FString Array(const TArray<FString>& EncodedValues)
	{
		FString Result = TEXT("[");
		for (int32 Index = 0; Index < EncodedValues.Num(); ++Index)
		{
			if (Index > 0)
			{
				Result.AppendChar(TEXT(','));
			}
			Result.Append(EncodedValues[Index]);
		}
		Result.AppendChar(TEXT(']'));
		return Result;
	}
}

File: sdk/unity/Runtime/ChillJson.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Text;

namespace Chill.Unity
{
    internal static class ChillJson
    {
        internal static string Quote(string value)
        {
            if (value == null)
            {
                return "null";
            }

            var result = new StringBuilder(value.Length + 8);
            result.Append('"');
            for (int index = 0; index < value.Length; index += 1)
            {
                char current = value[index];
                switch (current)
                {
                    case '"':
                        result.Append("\\\"");
                        break;
                    case '\\':
                        result.Append("\\\\");
                        break;
                    case '\b':
                        result.Append("\\b");
                        break;
                    case '\f':
                        result.Append("\\f");
                        break;
                    case '\n':
                        result.Append("\\n");
                        break;
                    case '\r':
                        result.Append("\\r");
                        break;
                    case '\t':
                        result.Append("\\t");
                        break;
                    default:
                        if (current < 0x20)
                        {
                            result.Append("\\u");
                            result.Append(((int)current).ToString("x4", CultureInfo.InvariantCulture));
                        }
                        else
                        {
                            result.Append(current);
                        }
                        break;
                }
            }
            result.Append('"');
            return result.ToString();
        }

        internal static string StringValue(string value)
        {
            return "{\"stringValue\":" + Quote(value) + "}";
        }

        internal static string BooleanValue(bool value)
        {
            return "{\"boolValue\":" + (value ? "true" : "false") + "}";
        }

        internal static string IntegerValue(long value)
        {
            return "{\"intValue\":" + Quote(value.ToString(CultureInfo.InvariantCulture)) + "}";
        }

        internal static string IntegerValue(string value)
        {
            return "{\"intValue\":" + Quote(value) + "}";
        }

        internal static string DoubleValue(double value)
        {
            return "{\"doubleValue\":" + value.ToString("R", CultureInfo.InvariantCulture) + "}";
        }

        internal static string StringArrayValue(IReadOnlyList<string> values)
        {
            var result = new StringBuilder("{\"arrayValue\":{\"values\":[");
            for (int index = 0; index < values.Count; index += 1)
            {
                if (index > 0)
                {
                    result.Append(',');
                }
                result.Append(StringValue(values[index]));
            }
            result.Append("]}}");
            return result.ToString();
        }

        internal static string Attribute(string key, string anyValueJson)
        {
            return "{\"key\":" + Quote(key) + ",\"value\":" + anyValueJson + "}";
        }

        internal static string AnyValue(ChillValue value)
        {
            switch (value.Kind)
            {
                case ChillValueKind.String:
                    return StringValue((string)value.Value);
                case ChillValueKind.Boolean:
                    return BooleanValue((bool)value.Value);
                case ChillValueKind.Number:
                    return DoubleValue((double)value.Value);
                case ChillValueKind.Strings:
                    return StringArrayValue((string[])value.Value);
                case ChillValueKind.Booleans:
                    return BooleanArrayValue((bool[])value.Value);
                case ChillValueKind.Numbers:
                    return NumberArrayValue((double[])value.Value);
                default:
                    throw new InvalidOperationException("Unsupported Chill annotation value.");
            }
        }

        internal static string Array(IEnumerable<string> encodedValues)
        {
            var result = new StringBuilder("[");
            bool first = true;
            foreach (string encoded in encodedValues)
            {
                if (!first)
                {
                    result.Append(',');
                }
                first = false;
                result.Append(encoded);
            }
            result.Append(']');
            return result.ToString();
        }

        private static string BooleanArrayValue(IReadOnlyList<bool> values)
        {
            var result = new StringBuilder("{\"arrayValue\":{\"values\":[");
            for (int index = 0; index < values.Count; index += 1)
            {
                if (index > 0)
                {
                    result.Append(',');
                }
                result.Append(BooleanValue(values[index]));
            }
            result.Append("]}}");
            return result.ToString();
        }

        private static string NumberArrayValue(IReadOnlyList<double> values)
        {
            var result = new StringBuilder("{\"arrayValue\":{\"values\":[");
            for (int index = 0; index < values.Count; index += 1)
            {
                if (index > 0)
                {
                    result.Append(',');
                }
                result.Append(DoubleValue(values[index]));
            }
            result.Append("]}}");
            return result.ToString();
        }
    }
}
